<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from src/robox/description/robox.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="robox">
  <ros2_control name="diffbot_ros2_control" type="system">
    <hardware>
      <plugin>diffdrive_arduino/DiffDriveArduinoHardware</plugin>
      <param name="left_wheel_name">left_wheel_joint</param>
      <param name="right_wheel_name">right_wheel_joint</param>
      <param name="loop_rate">30</param>
      <param name="device">/dev/ttyUSB1</param>
      <param name="baud_rate">57600</param>
      <param name="timeout_ms">1000</param>
      <param name="enc_counts_per_rev">330</param>
      <param name="pid_p">20</param>
      <param name="pid_d">12</param>
      <param name="pid_i">0</param>
      <param name="pid_o">50</param>
    </hardware>
    <joint name="left_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="right_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>
  <!-- Inertial macros for calculating inertia for spheres, boxes, and cylinders -->
  <!-- Using standard inertial formulas for each shape -->
  <!-- Materials for visualization -->
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  <material name="orange">
    <color rgba="1 0.3 0.1 1"/>
  </material>
  <material name="blue">
    <color rgba="0.2 0.2 1 1"/>
  </material>
  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>
  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>
  <!-- BASE LINK -->
  <link name="base_link">
    </link>
  <!-- BASE FOOTPRINT LINK -->
  <joint name="base_footprint_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_footprint"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <link name="base_footprint">
    </link>
  <!-- CHASSIS LINK -->
  <joint name="chassis_joint" type="fixed">
    <parent link="base_link"/>
    <child link="chassis"/>
    <origin xyz="-0.16499999999999998 0 -0.030000000000000002"/>
  </joint>
  <link name="chassis">
    <visual>
      <origin xyz="0.1225 0 0.0375"/>
      <geometry>
        <box size="0.245 0.145 0.075"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin xyz="0.1225 0 0.0375"/>
      <geometry>
        <box size="0.245 0.145 0.075"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.1225 0 0.0375"/>
      <mass value="1.3"/>
      <inertia ixx="0.0028870833333333335" ixy="0.0" ixz="0.0" iyy="0.007112083333333334" iyz="0.0" izz="0.008780416666666667"/>
    </inertial>
  </link>
  <!-- LEFT WHEEL LINK -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0.09949999999999999 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.027" radius="0.035"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.035"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.06"/>
      <inertia ixx="2.202e-05" ixy="0.0" ixz="0.0" iyy="2.202e-05" iyz="0.0" izz="3.6750000000000006e-05"/>
    </inertial>
  </link>
  <!-- RIGHT WHEEL LINK -->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 -0.09949999999999999 0"/>
    <axis xyz="0 0 -1"/>
  </joint>
  <link name="right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.027" radius="0.035"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.035"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.06"/>
      <inertia ixx="2.202e-05" ixy="0.0" ixz="0.0" iyy="2.202e-05" iyz="0.0" izz="3.6750000000000006e-05"/>
    </inertial>
  </link>
  <!-- CASTER WHEEL LINK -->
  <joint name="caster_wheel_joint" type="fixed">
    <parent link="chassis"/>
    <child link="caster_wheel"/>
    <origin xyz="0.03 0 -8.673617379884035e-19"/>
  </joint>
  <link name="caster_wheel">
    <visual>
      <geometry>
        <sphere radius="0.005"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.005"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.01"/>
      <inertia ixx="1.0000000000000001e-07" ixy="0.0" ixz="0.0" iyy="1.0000000000000001e-07" iyz="0.0" izz="1.0000000000000001e-07"/>
    </inertial>
  </link>
  <joint name="laser_joint" type="fixed">
    <parent link="chassis"/>
    <child link="laser_frame"/>
    <origin rpy="0 0 0" xyz="0.185 0 0.12"/>
    <!-- Adjusted according to ground to LiDAR top: 14.5cm -->
  </joint>
  <link name="laser_frame">
    <visual>
      <geometry>
        <cylinder length="0.03" radius="0.04"/>
        <!-- Approximate size of LiDAR -->
      </geometry>
      <material name="black"/>
    </visual>
    <visual>
      <origin xyz="0 0 -0.05"/>
      <geometry>
        <cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.04" radius="0.04"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.125"/>
      <inertia ixx="6.666666666666666e-05" ixy="0.0" ixz="0.0" iyy="6.666666666666666e-05" iyz="0.0" izz="0.0001"/>
    </inertial>
  </link>
</robot>
